pyccl/pkresponse.py

```diff
@@ -0,0 +1,805 @@
+    hmc = halos.HMCalculator(
+        mass_function=hmf,
+        halo_bias=hbf,
+        mass_def=mass_def,
+        log10M_min=log10M_min,
+        log10M_max=log10M_max,
+        nM=nM,
+    )
+
+    logM = hmc._lmass
+    M = hmc._mass
+    Mh = M * h
+    dlogM = logM[1] - logM[0]
+
+    for ia, aa in enumerate(a_arr):
+        z = 1.0 / aa - 1
+        hmc._get_ingredients(cosmo, aa, get_bf=True)
+
+        _darkemu_set_cosmology(emu, cosmo)
+        for m in range(nM):
+            hmc_m = halos.HMCalculator(
+                mass_function=hmf,
+                halo_bias=hbf,
+                mass_def=mass_def,
+                log10M_min=logM[m],
+                log10M_max=17.0,
+            )
+            hmc_m._get_ingredients(cosmo, aa, get_bf=True)
+
+            Pth[m] = emu.get_phm_massthreshold(k_emu, Mh[m], z) * (1 / h) ** 3
+            Pbin[m] = emu.get_phm_mass(k_emu, Mh[m], z) * (1 / h) ** 3
+            nths[m] = emu.mass_to_dens(Mh[m], z) * h**3
+
+            array_2 = np.ones(len(hmc_m._mass))
+            array_2[..., 0] = 0
+            b1_th_tink[m] = hmc_m._integrate_over_mbf(
+                array_2
+            ) / hmc_m._integrate_over_mf(array_2)
+
+        _darkemu_set_cosmology(emu, cosmo_hp)
+        for m in range(nM):
+            Pnth_hp[m] = (
+                emu.get_phm(
+                    k_emu * (h / hp), np.log10(nths[m] * (1 / hp) ** 3), z
+                )
+                * (1 / hp) ** 3
+            )
+
+        _darkemu_set_cosmology(emu, cosmo_hm)
+        for m in range(nM):
+            Pnth_hm[m] = (
+                emu.get_phm(
+                    k_emu * (h / hm), np.log10(nths[m] * (1 / hm) ** 3), z
+                )
+                * (1 / hm) ** 3
+            )
+
+        Nc = prof_hod._Nc(M, aa)
+        Ns = prof_hod._Ns(M, aa)
+        fc = prof_hod._fc(aa)
+        Ng = Nc * (fc + Ns)
+        logMps = logM + dlogM
+        logMms = logM - dlogM
+
+        prof_Mp = prof_hod.fourier(cosmo, k_use, (10**logMps), aa)
+        prof_Mm = prof_hod.fourier(cosmo, k_use, (10**logMms), aa)
+
+        dprof_dlogM = (prof_Mp - prof_Mm) / (2 * dlogM)
+
+        nth_mat = np.tile(nths, (len(k_use), 1)).transpose()
+
+        # Eq. 18
+        ng = hmc._integrate_over_mf(Ng)
+
+        # Eq. 17
+        bgE = hmc._integrate_over_mbf(Ng) / ng
+
+        # Eq. 19
+        bgE2 = hmc._integrate_over_mf(Ng * _b2H17(hmc._bf)) / ng
+
+        bgL = bgE - 1
+
+        b1L_th_mat = np.tile(b1_th_tink - 1, (len(k_emu), 1)).transpose()
+
+        dPhm_db_nfix = (
+            (26.0 / 21.0)
+            * np.log(np.array(Pnth_hp) / np.array(Pnth_hm))
+            * np.array(Pth)
+            / (2 * (np.log(Dp[ia]) - np.log(Dm[ia])))
+        )  # Mpc^3
+
+        dnP_hm_db_emu = nth_mat * (dPhm_db_nfix + b1L_th_mat * np.array(Pbin))
+
+        # Eq. A2
+        nP = nth_mat * np.array(Pth)
+
+        # Eq. A7
+        Pgm = integrate.romb(dprof_dlogM * nP, dx=dlogM, axis=0) / ng
```

Comment:
I think this could be done with current halo calculator, @damonge?

I've commented those that I think might use current CCL integrators